File: src/levels/meta/level22.ts
import type { LevelMeta } from '../../lib/types';

const meta: LevelMeta = {
  id: 22,
  slug: 'the-strategy-factory',
  title: 'The Strategy Factory',
  tier: 'QUANT GAUNTLET',
  tagline: '100 backtests walk into a bar. the bar is where they should stay.',
  brief: String.raw`A research team proudly delivers **100 strategy backtests** — two years of daily equity curves, \`s001\` through \`s100\`. Several look spectacular. The team wants capital allocated to the best one.

Here is what they didn't tell you: **99 of these are pure noise.** Coin flips with a P&L axis. Exactly one has a real (modest) edge.

The best of 99 coin-flip Sharpe ratios is *expected* to be around 1.8 annualized. Eyeballing the leaderboard is worthless. Your deliverables: (1) the index of the real strategy, and (2) the **familywise error rate** $\alpha$ at which your identification actually holds up — too small is overclaiming, too large isn't a test. On some other draw of this world, no answer would be defensible; part of the job is knowing whether you're in that world.`,
  recover: 'Recover the real strategy’s index and state a defensible FWER α.',
  submitNote:
    'α is graded as a range: it must be at least the achieved p-value of your best-of-100 test (no overclaiming) and at most 0.10.',
  starterCode: String.raw`# GROUND TRUTH · CASE 22 — The Strategy Factory
names = sorted(c for c in data.columns if c != "t")
E = data[names].to_numpy()
R = np.diff(np.log(E), axis=0)          # daily returns, 504 x 100
sharpes = R.mean(0) / R.std(0, ddof=1) * np.sqrt(252)

order = np.argsort(sharpes)[::-1]
for i in order[:8]:
    print(f"{names[i]}  Sharpe {sharpes[i]:+.2f}")
# tempting, right? one question before you allocate:
# how many strategies did the factory test?
`,
  hints: {
    nudge: String.raw`The question is never "is this Sharpe big?" — it's "is this Sharpe big *given that I picked the maximum of 100 tries*?" You need the distribution of $\max_{100} \widehat{SR}_{\text{null}}$, not of $\widehat{SR}$.`,
    method: String.raw`White's reality check by simulation: a null strategy's annualized Sharpe over $T$ days is $\approx \mathcal{N}(0, 252/T)$ — std $\approx 0.707$ here. Simulate (or bootstrap from the curves themselves) many panels of 100 null Sharpes, record the max each time. Your achieved FWER p-value is $P(\max_{100} \ge \widehat{SR}_{\text{best,obs}})$. If that p is small (it is, on this seed — but check!), the argmax is your index and any stated $\alpha \ge p$ (up to 0.10) is defensible.`,
    derivation: String.raw`Extreme-value arithmetic: for $n$ independent $\mathcal{N}(0, \sigma_S^2)$ Sharpes, $\mathbb{E}[\max] \approx \sigma_S\big((1-\gamma)\Phi^{-1}(1 - \tfrac1n) + \gamma\,\Phi^{-1}(1 - \tfrac{1}{ne})\big)$ (Bailey–López de Prado's Expected Maximum Sharpe, $\gamma \approx 0.5772$). With $\sigma_S = 0.707, n = 100$: $\mathbb{E}[\max] \approx 1.77$, and the 95th percentile of the max sits near $2.4$. This is the **Deflated Sharpe Ratio** idea: a backtest's Sharpe must clear the *expected best of your trial count*, not zero. Selection bias isn't a small correction — it's the whole first-order effect.

FWER logic: testing 100 hypotheses at per-test $\alpha = 0.05$ yields $\approx 5$ false discoveries per panel — someone always looks brilliant. Controlling the *familywise* rate means calibrating against the max-statistic distribution (equivalent to Bonferroni/Šidák at this independence level: per-test $p \cdot 100 \lesssim \alpha$).

The uncomfortable epistemology, which is the real cargo of this level: the defensible $\alpha$ is a property of **the data plus your honesty about the search**, not of your conviction. If the team had run 10,000 backtests instead of 100, $\mathbb{E}[\max]$ rises to $\approx 2.7$ and this same winner would be unidentifiable — *identical returns, different truth*, because the trial count is part of the evidence. This is why "how many things did you try?" is the first question a good allocator asks, and why unreported trials are the finance version of $p$-hacking. Harvey–Liu–Zhu estimate most published factors fail exactly this test.`,
  },
  xLabel: 't (days)',
  yLabel: 'equity',
};

export default meta;

File: datagen/generators/level22.py
"""Level 22 — The Strategy Factory. 100 backtests; 99 noise, 1 real edge.

Multiple-testing: the best of 99 coin-flip strategies looks brilliant.
Intended method: White's reality check / deflated Sharpe — compare the best
observed Sharpe against the distribution of the MAX of 100 null Sharpes.

Grading: the index of the real strategy (integer) AND a defensible
familywise error rate: stated alpha must be >= the achieved FWER p-value
(no overclaiming) and <= 0.10 (no rubber stamps). The seed is chosen so the
real strategy is identifiable — the test asserts it; on unlucky seeds the
generator itself fails loudly and a new seed must be chosen.
"""

import numpy as np

from datagen.common import make_pack, field, range_field, rng, truth_block

SEED = 3234  # chosen: true strategy is argmax, p_fwer ~ 0.008, and a noise impostor reaches Sharpe 2.1
N_STRATS = 100
T_DAYS = 504  # 2 years
SIGMA_D = 0.01
TRUE_IDX = 37          # 1-based index of the real strategy
SHARPE_TRUE = 3.0      # annualized
MU_D = SHARPE_TRUE * SIGMA_D / np.sqrt(252.0)


def simulate(g: np.random.Generator) -> np.ndarray:
    """Daily returns matrix (T x N). Column TRUE_IDX-1 has the edge."""
    R = g.normal(0.0, SIGMA_D, (T_DAYS, N_STRATS))
    R[:, TRUE_IDX - 1] += MU_D
    return R


def annual_sharpes(R: np.ndarray) -> np.ndarray:
    return R.mean(axis=0) / R.std(axis=0, ddof=1) * np.sqrt(252.0)


def max_null_sharpe_dist(n_sims: int = 20000) -> np.ndarray:
    """Distribution of max over 100 null annualized Sharpes (exact simulation)."""
    g = rng(202_200)
    # Sharpe of T iid N(0, s^2) days: mean/std * sqrt(252); simulate directly
    # in sufficient-statistic form for speed.
    z = g.normal(0.0, 1.0, (n_sims, N_STRATS))  # mean part
    chi = g.chisquare(T_DAYS - 1, (n_sims, N_STRATS))
    sharpe = (z / np.sqrt(T_DAYS)) / np.sqrt(chi / (T_DAYS - 1)) * np.sqrt(252.0)
    return sharpe.max(axis=1)


def build() -> dict:
    g = rng(SEED)
    R = simulate(g)
    sharpes = annual_sharpes(R)

    # the real strategy must be THE argmax on this seed, else change the seed
    assert int(np.argmax(sharpes)) == TRUE_IDX - 1, (
        f"seed {SEED}: real strategy (S={sharpes[TRUE_IDX-1]:.2f}) beaten by "
        f"strategy {int(np.argmax(sharpes))+1} (S={sharpes.max():.2f}) — change the seed"
    )

    # achieved FWER p-value: P(max of 100 null Sharpes >= observed best)
    null_max = max_null_sharpe_dist()
    p_fwer = float(np.mean(null_max >= sharpes[TRUE_IDX - 1]))
    assert p_fwer < 0.05, f"seed {SEED}: not identifiable at 5% (p={p_fwer:.3f})"
    alpha_lo = max(p_fwer, 0.001)

    equity = np.exp(np.cumsum(np.vstack([np.zeros(N_STRATS), R]), axis=0))
    t = np.arange(T_DAYS + 1, dtype=float)
    columns = {"t": t}
    for j in range(N_STRATS):
        columns[f"s{j+1:03d}"] = equity[:, j]

    return make_pack(
        id=22,
        seed=SEED,
        kind="panel",
        columns=columns,
        truth=truth_block(
            {
                "index": float(TRUE_IDX),
                "sharpe_true": SHARPE_TRUE,
                "p_fwer": p_fwer,
                "expected_max_null": float(np.median(null_max)),
            },
            r"99 \times \mathcal{N}(0, \sigma^2)\ \text{noise};\ \ 1\ \text{real edge at index } 37.\quad \mathbb{E}\big[\max_{99}\ \widehat{SR}_{\text{null}}\big] \approx 1.8",
        ),
        grading={
            "mode": "parameter",
            "fields": [
                field(
                    "index",
                    "index of the real strategy",
                    float(TRUE_IDX),
                    0.5,
                    integer=True,
                    help="1–100; or convince yourself none is identifiable — then check the derivation hint",
                ),
                range_field(
                    "alpha",
                    "your familywise error rate α",
                    alpha_lo,
                    0.10,
                    help="the FWER at which your identification is defensible: too small = overclaiming, too large = not a test",
                ),
            ],
        },
        stats_col="s001",  # never the true strategy: stats must not leak the answer
    )
